// @abaplint/runtime/build/src/types/numc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Numc = void 0;
class Numc {
    constructor(input) {
        this.length = (input === null || input === void 0 ? void 0 : input.length) ? input === null || input === void 0 ? void 0 : input.length : 1;
        this.qualifiedName = input === null || input === void 0 ? void 0 : input.qualifiedName;
        this.clear();
    }
    getQualifiedName() {
        return this.qualifiedName;
    }
    set(value) {
        if (typeof value === "number") {
            this.value = value.toString();
        }
        else if (typeof value === "string") {
            this.value = parseInt(value, 10) + "";
        }
        else {
            this.set(value.get());
            return;
        }
        if (this.value.length > this.length) {
            this.value = this.value.substr(this.value.length - this.length, this.length);
        }
        else {
            const pad = this.length - this.value.length;
            if (pad > 0) {
                this.value = "0".repeat(pad) + this.value;
            }
        }
        return this;
    }
    getLength() {
        return this.length;
    }
    clear() {
        this.value = "0".repeat(this.length);
    }
    get() {
        return this.value;
    }
    getOffset(_input) {
        throw "todo, runtime, numc getOffset()";
    }
}
exports.Numc = Numc;
