// @abaplint/runtime/build/src/types/packed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Packed = void 0;
const float_1 = require("./float");
class Packed {
    constructor(input) {
        this.value = 0;
        this.length = 666;
        if (input === null || input === void 0 ? void 0 : input.length) {
            this.length = input.length;
        }
        this.decimals = 0;
        if (input === null || input === void 0 ? void 0 : input.decimals) {
            this.decimals = input.decimals;
        }
        this.qualifiedName = input === null || input === void 0 ? void 0 : input.qualifiedName;
    }
    getQualifiedName() {
        return this.qualifiedName;
    }
    set(value) {
        if (typeof value === "number") {
            this.value = value;
        }
        else if (typeof value === "string") {
            this.value = parseFloat(value);
        }
        else if (value instanceof float_1.Float) {
            this.value = value.getRaw();
        }
        else {
            this.set(value.get());
        }
    }
    getLength() {
        return this.length;
    }
    getDecimals() {
        return this.decimals;
    }
    clear() {
        this.value = 0;
    }
    get() {
        return this.value;
    }
}
exports.Packed = Packed;
